### PR TITLE
Integrate the file globbing with Project.

### PR DESCRIPTION
**Pull request by troydai**
The functional tests are being reviewed in another PR, please take a look them first: https://github.com/aspnet/XRE/pull/1175

### CODE REVIEW

File: src/Microsoft.Framework.Runtime.Loader/Microsoft.Framework.Runtime.Loader.kproj [hunk 1]
```diff
@@ -14,4 +14,9 @@
     <SchemaVersion>2.0</SchemaVersion>
   </PropertyGroup>
   <Import Project="$(VSToolsPath)\AspNet\Microsoft.Web.AspNet.targets" Condition="'$(VSToolsPath)' != ''" />
+  <ProjectExtensions>
+    <VisualStudio>
+      <UserProperties project_1json__JSONSchema="http://www.asp.net/media/4878834/project.json" />
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Loader/Microsoft.Framework.Runtime.Loader.kproj]: ? Update your VS

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+using System.IO;
+using System.Linq;
+using Microsoft.Framework.FileSystemGlobbing;
+using Microsoft.Framework.FileSystemGlobbing.Abstractions;
+
+namespace Microsoft.Framework.Runtime.FileGlobbing
+{
+    public static class MatcherExtensions
+    {
+        public static void AddExcludePatterns(this Matcher self, params IEnumerable<string>[] excludePatternsGroups)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: self? This isn't python, it should be `matcher`

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 3]
```diff
@@ -857,5 +832,21 @@ private static string GetDirectoryName(string path)
             path = path.TrimEnd(Path.DirectorySeparatorChar);
             return path.Substring(Path.GetDirectoryName(path).Length).Trim(Path.DirectorySeparatorChar);
         }
+
+        private static void AddIncludePatterns(Matcher matcher, IEnumerable<string> patterns)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Project.cs]: Why do you need this as well as the extension methods?

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 4]
```diff
@@ -47,13 +56,7 @@ public Project()
...
         public string ProjectFilePath { get; private set; }
 
-        public string ProjectDirectory
-        {
-            get
-            {
-                return Path.GetDirectoryName(ProjectFilePath);
-            }
-        }
+        public string ProjectDirectory { get { return Path.GetDirectoryName(ProjectFilePath); } }
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Project.cs]: Why the change in formatting?

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Framework.FileSystemGlobbing;
+using Microsoft.Framework.FileSystemGlobbing.Abstractions;
+
+namespace Microsoft.Framework.Runtime.FileGlobbing
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: Keep the namespace the same as globbing. `Microsoft.Framework.FileSystemGlobbing`

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            foreach (var group in includePatternsGroups)
+            {
+                foreach (var pattern in group)
+                {
+                    self.AddInclude(pattern);
+                }
+            }
+        }
+
+        public static string[] GetResultsInFullPath(this Matcher self, string directoryPath)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: Why not make this `IEnumerable<string>`?

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            var relativePaths = self.Execute(new DirectoryInfoWrapper(new DirectoryInfo(directoryPath))).Files;
+            var result = relativePaths.Select(path => Path.Combine(directoryPath, path)).ToArray();
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: Call Path.GetFullPath?

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 3 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/Project.cs]: Nop, should be removed.

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs (hunk 5 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: Ok.

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs (hunk 7 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: Yup.

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 4 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/Project.cs]: Just to make the file more compact while working. I'll change it back.

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs (hunk 6 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: Yup.

File: src/Microsoft.Framework.Runtime.Loader/Microsoft.Framework.Runtime.Loader.kproj (hunk 1 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime.Loader/Microsoft.Framework.Runtime.Loader.kproj]: Will revert this.

File: src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs (hunk 2 above)
Comment (troydai) [inline src/Microsoft.Framework.Runtime/FileGlobbing/MatcherExtensions.cs]: ok

**Comment by troydai** [comment]
Need more work.
